### ISSUE TITLE
[Slightly off topic] drgn reading different values from /proc/kcore w/ CTF and ptrace

### ISSUE THREAD

**Issue by brenns10** [osandov/drgn#448]
To preface this, it's almost certainly not a drgn bug which makes it a bit off topic. If I had to guess, it's a `/proc/kcore` bug. But it's a fascinating odd behavior which I'm observing and wanted to document.

---

My CI testing system for drgn-tools involves booting some VMs into different Oracle Linux versions and running the drgn-tools tests on each. The earliest version I test is Oracle Linux 7, with a kernel that's based on the 4.1 series (UEK 4). So this is real old stuff. I observed a test failure when I was running the CI tests with CTF debuginfo rather than DWARF. I tracked it down to a specific kernel memory address that drgn is reading, which is giving a certain value for the drgn running w/ CTF, and a different value for DWARF.

```python
# with CTF:
>>> hex(prog.read_word(0xffff88013fc17b18))
'0xffff88013a452e10'

# with DWARF:
>>> hex(prog.read_word(0xffff88013fc17b18))
'0xffff88013fc17b18'
```

That's already strange, because no type information _should_ be used for the memory reader, so there should be no difference.

But I figured that it was possible that these were somehow turning into different `/proc/kcore` read requests. So I attached strace to both drgn processes and re-ran the same line of code. What's wild is that once I attached with `strace`, the drgn task running with CTF "magically" gets the correct value. The task with DWARF doesn't change its return value. Here's the strace output:
```
[pid 12451] pread64(3, "\30{\301?\1\210\377\377", 8, 8801457642264) = 8
 | 00000  18 7b c1 3f 01 88 ff ff                           .{.?....         |
[pid 12390] pread64(3, "\30{\301?\1\210\377\377", 8, 8801457642264) = 8
 | 00000  18 7b c1 3f 01 88 ff ff                           .{.?....         |
```
And here's the drgn terminal output:
```python
# with CTF:
>>> hex(prog.read_word(0xffff88013fc17b18))
'0xffff88013fc17b18'

# with DWARF:
>>> hex(prog.read_word(0xffff88013fc17b18))
'0xffff88013fc17b18'
```

So that's pretty crazy. Finally, I decided that I could cut drgn completely out of the loop here, by trying the read directly in Python. Here's the drgn process running with CTF:

```python
# with CTF, strace is attached:
>>> hex(struct.unpack("=Q", os.pread(3, 8, 8801457642264))[0])
'0xffff88013fc17b18'
# with CTF, now strace is detached:
>>> hex(struct.unpack("=Q", os.pread(3, 8, 8801457642264))[0])
'0xffff88013a452e10'
```

So basically the same `/proc/kcore` read gives two different values, depending on whether strace is attached to drgn, and somehow, depending on whether drgn is running with CTF (!!!).

I'm totally nerd sniped by this right now, but we will see how far I get. At the end of the day it's a pretty old kernel that's probably missing some /proc/kcore bug fixes, so that's probably going to be the explanation...

**Comment by brenns10** [comment]
Using `trace-cmd -p function_graph -l read_kcore` I didn't see any difference in behavior between the two calls. Even the timing was very similar.

The thing I neglected to mention initially was that this memory address is related to the workqueue code. I initially thought that it had a constant value (re-running the reads gave consistent values, after all). My guess is now that its value is regularly changing as `work_structs` get added and removed to the specific `worker_pool` object that it's related to, and I'm observing those changes. There must be something related to the specific task (and maybe ptrace) that influences what the value is at the time of the read.

This part of the kernel just moves too fast and I probably just got caught up in thinking that I could get reliable data from it from userspace.

**Comment by brenns10** [comment]
(Going to close this since it was a fun thing to write but definitely not helpful to keep open)
